File: Leetcode/532-k_diff_pairs_array.py
import collections

class Solution(object):
    def findPairs(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        Use Counter as a frequency map. Iterate over counter to handle 
        edge case of k == 0.
        """
        seen = collections.Counter(nums)
        total = 0
        for i in seen:
            if (i-k in seen and k > 0) or (k == 0 and seen[i-k]>1):
                total += 1
        return total

z = Solution()
nums = [1,2,3,4,5]
k = 1
nums = [1,3,1,5,4]
k = 0
# nums = [1,2,4,4,3,3,0,9,2,3]
# k = 3
# nums = [-1,-2,-3]
# k = 1
print(z.findPairs(nums, k))

File: Leetcode/532-k_diff_pairs_array.cpp
#include <iostream>
#include <vector>
#include <map>
#include <unordered_set>

using namespace std;

class Solution {
public:
    int findPairs(vector<int>& nums, int k) {
        unordered_set<int> recorded;
        map<int, int> seen;
        for (auto& num:nums) {
          seen[num]++;
        }
        int res = 0;
        for (auto& p:seen) {
          if ((k == 0 && p.second > 1) || (k > 0 && seen.count(p.first+k)>0)) {
            res++;
          }
        }
        return res;
    }
};

int main() {
    Solution solution;
    vector<int> nums = {3, 1, 4, 1, 5};
    int k = 4;

    cout << solution.findPairs(nums, k) << endl;
    return 0;
}